src/main/java/org/ethereum/beacon/discovery/task/RecursiveLookupTasks.java

```diff
@@ -33,36 +32,37 @@ public RecursiveLookupTasks(
     this.discoveryManager = discoveryManager;
     this.scheduler = scheduler;
     this.taskTimeouts =
-        new ExpirationScheduler<>(timeout.get(ChronoUnit.NANOS), TimeUnit.NANOSECONDS);
+        new ExpirationScheduler<>(timeout.get(ChronoUnit.SECONDS), TimeUnit.SECONDS);
   }
 
-  public void add(NodeRecordInfo nodeRecordInfo, Runnable successCallback, Runnable failCallback) {
-    synchronized (this) {
-      if (currentTasks.contains(nodeRecordInfo.getNode().getNodeId())) {
-        return;
-      }
-      currentTasks.add(nodeRecordInfo.getNode().getNodeId());
+  public CompletableFuture<Void> add(NodeRecord nodeRecord, int distance) {
+    if (!currentTasks.add(nodeRecord.getNodeId())) {
+      return CompletableFuture.failedFuture(new IllegalStateException("Already querying node"));
     }
 
+    final CompletableFuture<Void> result = new CompletableFuture<>();
     scheduler.execute(
         () -> {
-          CompletableFuture<Void> retry =
-              discoveryManager.findNodes(nodeRecordInfo.getNode(), DEFAULT_DISTANCE);
-          taskTimeouts.put(
-              nodeRecordInfo.getNode().getNodeId(),
-              () ->
-                  retry.completeExceptionally(
-                      new RuntimeException("Timeout for node recursive lookup task")));
+          CompletableFuture<Void> retry = discoveryManager.findNodes(nodeRecord, distance);
```

Comment:
Am I reading this wrong?  This doesn't look like a retry:
```suggestion
          CompletableFuture<Void> findNodesFuture = discoveryManager.findNodes(nodeRecord, distance);
```

Reply:
That is indeed not a retry but for some reason that's what it was called.  Somewhere in the changesets I've been wrangling I've renamed it but apparently it didn't make it to this PR. Updated here to make sure it doesn't get forgotten.